design_documents/cairo_deployment_scripts.md

```diff
@@ -0,0 +1,335 @@
+  let user = create_account("user1", &provider);
+  let prefund_account = invoke("0x123", "deposit", [1234, user.address], &prefunder);
+  deploy_account("user1", &provider);
+  get_account("user1", &provider)?
+}
+
+fn main() {
+  ler provider = get_provider("http://127.0.0.1:5050/rpc")?;
+  let user = make_account(provider);
+
+  let declared_contract = declare("mycontract", &user, "/path/to/Scarb.toml");
+  let contract = deploy(&declared_contract.class_hash, [], &user);
+
+  let dispatcher = DispatcherMyContract { contract.contract_address };
+
+  skip_if_done(|| {
+        dispatcher.increase_balance(5);
+    });
+  let get_balance = skip_if_done(|| {
+        dispatcher.get_balance();
+    });
+
+  let called_balance = call(&contract.contract_address, "get_balance", [], "latest"); 
+}
+```
+
+### example state file
+The state file by default should be written to the current working directory, with a name `<script file name>.state.json`.
+Its schema could look like this:
+
+```json
+{
+  "create_account": {
+    "arguments": {
+      "name": "whatever",
+      (...)
+    },
+    "output": {
+      "address": "0x123",
+      "max_fee": "0x321",
+    },
+    "status": "accepted",
+    "timestamp": (...),
+  },
+  "prefund_account": {
+    "arguments": {
+      ...
+    },
+    "output": {
+      "error": "..."
+    }
+  },
+  "status": "rejected",
+  "timestamp": (...),
+}
+```
+
+Having this, we could allow users to `--retries` or `--verify` transactions if something unexpected happens,
+some transactions were rejected because of faulty script, too low user eth balance etc.
+
+If we were to implement dry run mode that runs by default, the state should only be written to a file when we pass `--broadcast`
+flag; otherwise the output should be directed to stdout.
+
+The file should be used to see, if given action/transaction was already taken - it should take under account the function
+name as well as its parameters, to distinguish functions from one another (eg. user creates two accounts, we should check
+state file records, one by one, and check if a function `account new` was already invoked with specified parameter `username` 
+(and others, if supplied)).
+
+### Miscellaneous
+
+There would be a new subcommand `script` that would be invoked with required parameter - a path to script file, with
+`*.s.cairo` extension. Proposed interface:
+
+- `--gas-multiplier` - relative percentage by which to multiply add gas estimates (perhaps passed directly to sncast)
+- `--broadcast` - broadcasts the transactions
+- `--no-state-file` - do not write to state file
+- `--state-file` - specify path to custom state file
+- `--log-file` - output logs to a file
+- `--quiet` - do not output logs to stdout
+- `--slow` - makes sure a transaction is sent, only after its previous one has been confirmed and succeeded (possible reuse of `--wait` flag)
+- `--verify` - find a matching broadcast in state file, and try to verify transactions
```

Comment:
Can you explain this flag a bit more in depth? What do you mean by verify?

Reply:
Basically - for each transaction found in state file, we get a receipt and see if it exists, is accepted etc.

Reply:
Okay, so the default run would not compare anything agains the state file? Or would it just check if the transaction was sent without verifying it passed?

Reply:
If there is a state file we will always check it, and based on what's inside we'll execute functions or not. So for example:

```
create_account()
deploy_account()
invoke_something()
```

say we run the script that does those 3 things. For each of them we check the state file, and if it was already done (and suceeded!) we skip it. If it is not, we execute it and write to state file.

A `--verify` flag would add to this flow - for each of them functions we check the state file, and if it was already done we get its tx hash and check against the chain, to see if all out information from state file matches the status on chain (eg, is the status changed etc). If given function was not executed yet, we do it and write to state file.

Hope that clarifies it :)

Reply:
Makes sense now, would be good to include part of this explanation in the document.